testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs

```diff
@@ -0,0 +1,480 @@
+                        }
+                    }
+                    Err(_) => bail!("Waiting for pod {} to be deleted..", name),
+                }
+            })
+        })
+        .await
+        .map_err(|e| format_err!("Failed to delete pod {}: {:?}", name, e))
+    }
+}
+
+#[async_trait]
+impl ClusterSwarm for ClusterSwarmKube {
+    async fn upsert_validator(
+        &self,
+        index: u32,
+        num_validators: u32,
+        num_fullnodes: u32,
+        image_tag: &str,
+        delete_data: bool,
+    ) -> Result<()> {
+        debug!("upsert_validator");
+        let pod_name = format!("validator-{}", index);
+        let pod_api = Api::v1Pod(self.client.clone()).within(DEFAULT_NAMESPACE);
+        if pod_api.get(&pod_name).await.is_ok() {
+            debug!("Found pod, deleting it");
+            self.delete_pod(&pod_name).await?;
+        }
+        let node_name = if let Some(instance) = self.validator_to_node.lock().await.get(&index) {
+            instance.k8s_node().to_string()
+        } else {
+            "".to_string()
+        };
+        debug!("Creating pod {} on node {:?}", pod_name, node_name);
+        match pod_api
+            .create(
+                &PostParams::default(),
+                self.validator_spec(
+                    index,
+                    num_validators,
+                    num_fullnodes,
+                    &node_name,
+                    image_tag,
+                    delete_data,
+                ),
+            )
+            .await
+        {
+            Ok(o) => {
+                debug!("Created {}", o.metadata.name);
+            }
+            Err(e) => bail!("Failed to create pod {} : {}", pod_name, e),
+        }
+        if node_name.is_empty() {
+            let (node_name, pod_ip) = self.get_pod_node_and_ip(&pod_name).await?;
+            let ac_port = AdmissionControlConfig::default().address.port() as u32;
+            let instance = Instance::new_k8s(pod_name, pod_ip, ac_port, node_name);
+            self.validator_to_node.lock().await.insert(index, instance);
+        }
+        Ok(())
+    }
+
+    async fn delete_validator(&self, index: u32) -> Result<()> {
+        debug!("delete_validator");
+        let pod_name = format!("validator-{}", index);
+        self.delete_pod(&pod_name).await
+    }
+
+    async fn upsert_fullnode(
+        &self,
+        fullnode_index: u32,
+        num_fullnodes_per_validator: u32,
+        validator_index: u32,
+        num_validators: u32,
+        image_tag: &str,
+        delete_data: bool,
+    ) -> Result<()> {
+        debug!("upsert_fullnode");
+        let pod_name = format!("fullnode-{}-{}", validator_index, fullnode_index);
+        let pod_api = Api::v1Pod(self.client.clone()).within(DEFAULT_NAMESPACE);
+        if pod_api.get(&pod_name).await.is_ok() {
+            debug!("Found pod, deleting it");
+            self.delete_pod(&pod_name).await?;
+        }
+        let node_name = if let Some(instance) = self
```

Comment:
I feel like in many cases for some reason you use `""` instead of Option, this is bad pattern, can you update those places so that you use None vs Some, instead of ""?

Reply:
Updated

Reply:
Did you push updated change? I still see `"".to_string()` here.